pkg/config/tlscfg/reload_test.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlscfg
```

Comment:
Cases to test:

1. cert has been updated -- no new cert should be added to the pool, the existing cert should be updated
1. cert has been revoked -- it should still exist at the cert pool, but now with the new state (same as updated, I believe)
1. cert has different subject now -- cert for the old subject has to be removed
1. cert is a link to a link (similar to how config maps work in Kubernetes)
1. cert file has been removed -- what should it do? keep an old, potentially invalid cert? disable TLS? kill the server?

Reply:
I would keep this simpler. 

Instead of watching the files/certs we can _always_ load them when the `GetClientCertificate`/`GetCertificate` is being called. However the CA pull is different - we need to watch the files/certs bc there bc there is no callback like `GetCertificate`.

Reply:
> we can always load them when the GetClientCertificate/GetCertificate is being called

Doesn't it mean that you'd be doing a file system IO for each incoming request? Wouldn't this have a bad impact on performance?

Reply:
The get certificate functions are not called per each request. Based on my observation the   `GetClientCertificate`/ `GetClientCertificate` were called if once if the cert has changed on the opposite side.

Reply:
Interesting, judging by the godoc on `GetClientCertificate`, I would expect it to be called for each new connection, which is potentially every new request when doing a `curl`. Have you tested this only in the context of ES, or did you test this logic also in other TLS places, like the collector's HTTP port?

Reply:
You are right the `GetClientCertificate` is called per connection. I have tested this with grpc reporter. The cert has to be cached to avoid IO per request. 

In case of ES it seems that the connection was reused as the `GetClientCertificate` wasn't called always.